### PR TITLE
GCP: Least privilege policy implementation

### LINKED ISSUE #196: Least Priviledge - GCP
Implement least privilege resource permissions so that developers only have the minimal set of permissions required to interact with resources they have defined.

### PR DESCRIPTION
**Pull request by tjholm**
Firestore has no least privileged support at the collection level, so we'll need to document that as a caveat for this provider.

### CODE REVIEW

File: pkg/provider/pulumi/gcp/cloudrunner.go [hunk 1]
```diff
@@ -56,40 +55,31 @@ func (g *gcpProvider) newCloudRunner(ctx *pulumi.Context, name string, args *Clo
...
+	// // Give secret accessor permissions
+	// _, err = projects.NewIAMMember(ctx, name+"-secret-access", &projects.IAMMemberArgs{
+	// 	Role: pulumi.String("roles/secretmanager.secretAccessor"),
+	// 	// Get the cloudrun service account email
+	// 	Member:  pulumi.Sprintf("serviceAccount:%s", sa.Email),
+	// 	Project: pulumi.String(args.ProjectId),
+	// }, append(opts, pulumi.Parent(res))...)
+	// if err != nil {
+	// 	return nil, errors.WithMessage(err, "iam member "+name)
+	// }
```
Comment (asalkeld) [inline pkg/provider/pulumi/gcp/cloudrunner.go]: not crazy about the commented out code

Comment (asalkeld) [inline pkg/provider/pulumi/gcp/cloudrunner.go]: not crazy about the commented out code

File: pkg/provider/pulumi/gcp/policy.go [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+
+			switch resource.Type {
+			case v1.ResourceType_Bucket:
+				b := args.Resources.Buckets[resource.Name]
+
+				storage.NewBucketIAMMember(ctx, memberName, &storage.BucketIAMMemberArgs{
+					Bucket: b.Name,
+					Member: memberId,
+					Role:   rolePolicy.Name,
+				}, pulumi.Parent(res))
```
Comment (asalkeld) [inline pkg/provider/pulumi/gcp/policy.go:229]:
```suggestion
				}, append(opts, pulumi.Parent(res))...)
```
as with the others below

Comment (asalkeld) [inline pkg/provider/pulumi/gcp/policy.go:229]:
```suggestion
				}, append(opts, pulumi.Parent(res))...)
```
as with the others below

File: pkg/provider/pulumi/gcp/policy.go [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+
+	if err != nil {
+		return nil, err
+	}
+
+	for _, principal := range args.Policy.Principals {
+		sa := args.Principals[v1.ResourceType_Function][principal.Name]
+		name := newCustomRoleName(principal.Name)
+
+		for _, resource := range args.Policy.Resources {
```
Comment (asalkeld) [inline pkg/provider/pulumi/gcp/policy.go:217]: should we check args.Policy.Resources for nil?

Comment (asalkeld) [inline pkg/provider/pulumi/gcp/policy.go:217]: should we check args.Policy.Resources for nil?

File: pkg/provider/pulumi/gcp/policy.go (hunk 2 above)
Comment (tjholm) [reply pkg/provider/pulumi/gcp/policy.go:229]: I'm not sure we want to propogate resource options to all child resources, but apply them to the parent resource only.

Comment (tjholm) [reply pkg/provider/pulumi/gcp/policy.go:229]: resolving as duplicate comment

File: pkg/provider/pulumi/gcp/policy.go (hunk 3 above)
Comment (tjholm) [reply pkg/provider/pulumi/gcp/policy.go:217]: we should implement in the contracts and use generated validation classes, on the gRPC server when we're collecting resources so we can fail as soon as possible.

Comment (tjholm) [reply pkg/provider/pulumi/gcp/policy.go:217]: resolving as duplicate comment.

File: pkg/provider/pulumi/gcp/cloudrunner.go (hunk 1 above)
Comment (tjholm) [reply pkg/provider/pulumi/gcp/cloudrunner.go]: removed

Comment (tjholm) [reply pkg/provider/pulumi/gcp/cloudrunner.go]: removed

File: pkg/provider/pulumi/gcp/policy.go (hunk 2 above)
Comment (asalkeld) [reply pkg/provider/pulumi/gcp/policy.go:229]: ok, well some of the providers do this. so we should at least be consistent one way or the other.

**Comment by asalkeld** [comment]
@tjholm  needs a rebase, but otherwise lgtm

**Review by asalkeld — APPROVED** [review]
lgtm

**Pull request merged**